### PR TITLE
Math Test sanitising

### PR DESCRIPTION
**Pull request by baykaner**
- unified test types
- in some cases more types, or better tolerance is now used

### CODE REVIEW

File: cmake/BuildTools.cmake [hunk 1]
```diff
@@ -149,6 +149,7 @@ function (_internal_add_fetch_test
     target_link_libraries(${name} PRIVATE ${library} gmock gmock_main)
     target_include_directories(${name}
                                PRIVATE ${FETCH_ROOT_VENDOR_DIR}/googletest/googletest/include)
+    target_include_directories(${name} PRIVATE "${CMAKE_CURRENT_SOURCE_DIR}/${directory}")
```
Comment (evsmithx) [inline cmake/BuildTools.cmake:152]: I'm not sure what this does?

File: libs/math/tests/unit/activation_functions/sigmoid.cpp [hunk 2]
```diff
@@ -16,19 +16,23 @@
...
 //------------------------------------------------------------------------------
 
+#include "gtest/gtest.h"
 #include "math/activation_functions/sigmoid.hpp"
-#include "math/tensor.hpp"
+#include "test_types.hpp"
 
-#include "gtest/gtest.h"
+#include <iomanip>
+#include <iostream>
```
Comment (evsmithx) [inline libs/math/tests/unit/activation_functions/sigmoid.cpp]: Required imports?

File: libs/math/tests/unit/clustering/knn_test.cpp [hunk 3]
```diff
@@ -18,41 +18,41 @@
...
-  EXPECT_NEAR(double(output.at(1).second), double(4), 1e-4);
   EXPECT_EQ(output.at(2).first, SizeType(2));
-  EXPECT_NEAR(double(output.at(2).second), double(14.6969384), 1e-4);
   EXPECT_EQ(output.at(3).first, SizeType(3));
-  EXPECT_NEAR(double(output.at(3).second), double(16.6132477), 1e-4);
+
+  EXPECT_TRUE(output.at(0).second - double(2) < function_tolerance<DataType>());
+  EXPECT_TRUE(output.at(1).second - double(4) < function_tolerance<DataType>());
+  EXPECT_TRUE(output.at(2).second - double(14.6973) < function_tolerance<DataType>());
+  EXPECT_TRUE(output.at(3).second - double(16.6132477) < function_tolerance<DataType>());
```
Comment (evsmithx) [inline libs/math/tests/unit/clustering/knn_test.cpp]: Why has this changed from EXPECT_NEAR? The error messages from EXPECT_NEAR print both values and the tolerance which is helpful for debugging

File: libs/math/tests/unit/metrics/l2_norm.cpp [hunk 4]
```diff
@@ -16,37 +16,44 @@
...
+  test_array[2] = DataType(3);
+  test_array[3] = DataType(-4);
+  test_array[4] = DataType(5);
+  test_array[5] = DataType(-6);
+  test_array[6] = DataType(7);
+  test_array[7] = DataType(-8);
 
   // initialise to non-zero just to avoid correct value at initialisation
-  typename TypeParam::Type score(0);
+  DataType score(0);
```
Comment (evsmithx) [inline libs/math/tests/unit/metrics/l2_norm.cpp]: Incorrect comment?

File: libs/math/tests/unit/test_types.hpp [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+using FloatingTypes =
+    ::testing::Types<float, double, fetch::fixed_point::fp32_t, fetch::fixed_point::fp64_t>;
+using HighPrecisionFloatingTypes = ::testing::Types<float, double, fetch::fixed_point::fp64_t>;
+
+using IntAndFloatingTypes =
+    ::testing::Types<int32_t, int64_t, float, double, fetch::fixed_point::fp32_t,
+                     fetch::fixed_point::fp64_t>;
+using UnsignedIntAndFloatingTypes =
+    ::testing::Types<uint32_t, int32_t, uint64_t, int64_t, float, double,
+                     fetch::fixed_point::fp32_t, fetch::fixed_point::fp64_t>;
```
Comment (evsmithx) [inline libs/math/tests/unit/test_types.hpp:36]: Judging by the name I was surprised that this contained int32_t. Maybe we should call it FloatIntAndUIntTypes?

**Review by evsmithx — CHANGES_REQUESTED** [review]
It's a lot cleaner than it was before!

File: libs/math/tests/unit/distance/entropy_tests.cpp [hunk 6]
```diff
@@ -48,5 +43,9 @@ TYPED_TEST(EntropyTest, entropy)
   A.Set(SizeType{2}, DataType(0.3));
   A.Set(SizeType{3}, DataType(0.4));
```
Comment (MissingNO57) [inline libs/math/tests/unit/distance/entropy_tests.cpp:44]: I know these are just tests, but shouldn't it be static_cast<DataType>(0.3)?

File: libs/math/tests/unit/distance/euclidean_distance.cpp [hunk 7]
```diff
@@ -47,15 +43,15 @@ TYPED_TEST(EuclideanTest, simple_test)
...
   A.Set(SizeType{2}, DataType(3));
   A.Set(SizeType{3}, DataType(4));
-  EXPECT_NEAR(double(Euclidean(A, A)), 0, (double)function_tolerance<DataType>());
+  EXPECT_NEAR(double(distance::Euclidean(A, A)), 0, (double)function_tolerance<DataType>());
 
   ArrayType B = ArrayType(4);
   B.Set(SizeType{0}, static_cast<DataType>(1));
   B.Set(SizeType{1}, DataType(2));
   B.Set(SizeType{2}, DataType(3));
   B.Set(SizeType{3}, DataType(2));
```
Comment (MissingNO57) [inline libs/math/tests/unit/distance/euclidean_distance.cpp:52]: I know these are just tests, but shouldn't it be static_cast(0.3)?

File: libs/math/tests/unit/distance/euclidean_distance.cpp [hunk 8]
```diff
@@ -96,7 +92,7 @@ TYPED_TEST(EuclideanTest, matrix_euclidean_test)
   B.Set(SizeType{2}, SizeType{2}, DataType(-5));
   B.Set(SizeType{2}, SizeType{3}, DataType(-6));
```
Comment (MissingNO57) [inline libs/math/tests/unit/distance/euclidean_distance.cpp:93]: I know these are just tests, but shouldn't it be static_cast(-5)?

File: libs/math/tests/unit/distance/entropy_tests.cpp (hunk 6 above)
Comment (baykaner) [reply libs/math/tests/unit/distance/entropy_tests.cpp:44]: in this case it's a DataType constructor call, so it should be equivalent.

**Review by evsmithx — APPROVED** [review]

**Review by MissingNO57 — APPROVED** [review]
LGTM

**Pull request merged**